### DISCUSSION TITLE
"Error: An action could not be found at the URI" when attempting to reference a stand alone repo

### DISCUSSION THREAD

**Discussion by GavinCameron** [community/community#24593] [Apps, API and Webhooks]
<p>I created a repo in our org in order to create a ‘action container’ that could be generic enough to be used by any other repo’s in actions workflows.</p>
<p>When I attempt to use the action container I get an error message:</p>
<blockquote>
<p>Error: An action could not be found at the URI <code>Error: An action could not be found at the URI 'https://api.github.com/repos/actions/checkout/tarball/10c0fabc8ec....</code></p>
</blockquote>
<p>Here is a minimalist overview of first repo that I made. The purpose of this repo is to create an action container for SnowSQL for use by other repos within our org.</p>
<p>Repo <code>&lt;ourorg&gt;/my-ssql-container-action</code></p>
<ul>
<li>Dockerfile</li>
<li>action.yml</li>
<li>buildtime.sh</li>
</ul>
<p>Dockerfile:</p>
<pre><code class="lang-auto">### Draws from https://github.com/kurron/docker-snowsql/blob/b1b16ff88b36f9df58f692a69642e84f353c2d46/Dockerfile ###

FROM ubuntu:18.04

RUN apt-get update &amp;&amp; apt-get --assume-yes install -y curl

# Create non-root user
RUN groupadd --system snowflake --gid 444 &amp;&amp; \
useradd --uid 444 --system --gid snowflake --home-dir /home/snowflake --create-home --shell /sbin/nologin --comment "Docker image user" snowflake &amp;&amp; \
chown -R snowflake:snowflake /home/snowflake

# default to being in the user's home directory
WORKDIR /home/snowflake

ENV LC_ALL=C.UTF-8
ENV LANG=C.UTF-8
ENV VERSION 1.2.9
ENV SNOWSQL_DEST /usr/local/bin
ENV SNOWSQL_LOGIN_SHELL /home/snowflake/.bashrc

# grab the installation script
RUN curl -O https://sfc-repo.snowflakecomputing.com/snowsql/bootstrap/1.2/linux_x86_64/snowsql-1.2.9-linux_x86_64.bash

# Install the tool
## Add unzip first, avoid error
RUN apt-get install -y unzip vim
RUN bash snowsql-1.2.9-linux_x86_64.bash

# Run commands via bash script that could not be done directly due to gh actions permissions restrictions
COPY ./buildtime.sh /home/snowflake/buildtime.sh
ENTRYPOINT ["/home/snowflake/buildtime.sh"]
</code></pre>
<p>action.yml:</p>
<pre><code class="lang-auto"># action.yml

name: 'get data'
description: 'Get data using snowsql and save in a volume for rscripts to use'
inputs:
  PRIVATE_KEY_PATH:
    decsription: 'Path to the pipelines private ssql key'
    required: true
  SNOWSQL_PRIVATE_KEY_PASSPHRASE:
    description: 'Password to your ssql private key'
    required: true
  AWS_REGION:
    description: 'AWS_REGION'
    required: true
  AWS_ACCESS_KEY_ID:
    description: 'AWS key for S3'
    required: true
  AWS_SECRET_ACCESS_KEY:
    description: 'AWS_SECRET_ACCESS_KEY'
    required: true
  QUERY_FILE:
    description: 'path/to/the/sql/file/to/execute.sql'
    resuired: true
outputs:
  data-json:
    description: 'json output of the returned table'
runs:
  using: 'docker'
  image: 'Dockerfile'
  args:
    - ${{ inputs.PRIVATE_KEY_PATH }}
    - ${{ inputs.SNOWSQL_PRIVATE_KEY_PASSPHRASE }}
    - ${{ inputs.AWS_REGION }}
    - ${{ inputs.AWS_ACCESS_KEY_ID }}
    - ${{ inputs.AWS_SECRET_ACCESS_KEY }}
    - ${{ inputs.QUERY_FILE }}
</code></pre>
<p>buildtime.sh:</p>
<pre><code class="lang-auto">#!/bin/bash

# Switch to the non-root user
su - snowflake

# Run the SnowSQL client once, allowing it to auto-upgrade to the latest version.
# See https://docs.snowflake.com/en/user-guide/snowsql-install-config.html#label-understanding-auto-upgrades
snowsql -v

echo 'done!'
</code></pre>
<p>I tested these files on local using Docker, when I build I see the ‘done!’ message from buildtime.sh.</p>
<p>After confirming this works I pushed all 3 files to a repo within our org.</p>
<p>Then, in another repo I attempted to use this action within a workflow.</p>
<p>/some-repo-that-uses-snowsql:</p>
<ul>
<li>config <code># a config file for snowsql</code>
</li>
<li>training-data.sql <code># a sql file that the repo will run using the container action I created in </code>my-ssql-container-action</li>
<li>.github/workflows/main.yml</li>
</ul>
<p>Here’s the workflow file main.yml:</p>
<pre><code class="lang-auto">on: [push]

jobs:
  ssql_job:
    runs-on: ubuntu-latest
    name: Build ssql image
    steps:
      - name: Checkout # To use this repository's private action, you must check out the repository
        uses: actions/checkout@10c0fabc7ec358e5a3c4b0ab7fdce37c185ff34b
        with:
          repository: ourorg/my-ssql-container-action
          ref: main
      - name: Build
        uses: ./ # Uses an action in the root directory 
        with:
          config: ./config # the config file for snowsql
          query: ./training-data.sql
        env:
          SSQL_PRIVATE_KEY: ${{ secrets.SSQL_PRIVATE_KEY }}
          SNOWSQL_PRIVATE_KEY_PASSPHRASE: ${{ secrets.SNOWSQL_PRIVATE_KEY_PASSPHRASE }}
          AWS_REGION: 'us-east-1'
          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
          QUERY_FILE: 'training-data.sql'
</code></pre>
<p>Now, when I push to <code>&lt;ourorg&gt;/some-repo-that-uses-snowsql</code> I see the following under the actions tab:<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/2/b/2b6cfb27df4e4a30a0a5e0501f386c829ca7e276.png" data-download-href="/uploads/short-url/6ca1mZuDSAc5bAEGptNEId8YAZw.png?dl=1" title="Screenshot from 2021-08-24 14-16-02" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/52216341/181085540-b1f7d86d-f194-4303-977d-ba78cb7a79ed.png" alt="Screenshot from 2021-08-24 14-16-02" data-base62-sha1="6ca1mZuDSAc5bAEGptNEId8YAZw" width="690" height="186" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/2/b/2b6cfb27df4e4a30a0a5e0501f386c829ca7e276_2_690x186.png, https://aws1.discourse-cdn.com/github/optimized/3X/2/b/2b6cfb27df4e4a30a0a5e0501f386c829ca7e276_2_1035x279.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/2/b/2b6cfb27df4e4a30a0a5e0501f386c829ca7e276_2_1380x372.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/2/b/2b6cfb27df4e4a30a0a5e0501f386c829ca7e276_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screenshot from 2021-08-24 14-16-02</span><span class="informations">1465×395 38.6 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Am unsure what other information to provide.</p>
<p>How can I set up my workflow in order to reference my container action <code>&lt;ourorg&gt;/my-ssql-container-action</code> and use it within a workflow within <code>&lt;ourorg&gt;/some-repo-that-uses-snowsql</code>?</p>

**Comment by jsoref** [comment]
<p>The error message is just saying that it can’t retrieve a snapshot of actions/checkout for the specified sha.</p>
<p>I don’t know where this magic sha came from:</p>
<aside class="quote no-group" data-username="GavinCameron" data-post="1" data-topic="197435">
<div class="title">
<div class="quote-controls"></div>
 GavinCameron:</div>
<blockquote>
<pre><code class="lang-auto">        uses: actions/checkout@10c0fabc7ec358e5a3c4b0ab7fdce37c185ff34b
</code></pre>
</blockquote>
</aside>
<p>GitHub doesn’t believe in it. (Nor does Google!)</p>
<p>If you want to pin to the current value (from main), you could use:</p>
<pre><code class="lang-diff">-actions/checkout@10c0fabc7ec358e5a3c4b0ab7fdce37c185ff34b
+actions/checkout@afe4af09a72596f47d806ee5f8b2674ec07fdc73
</code></pre>
<p>The latest release appears to be <a href="https://github.com/actions/checkout/releases/tag/v2.3.4">5a4ac9002d0be2fb38bd78e4b4dbde5606d7042f</a></p>
<p>Personally, I’ve switched to using <a href="https://github.com/actions/checkout/commits/v2">@v2</a> – I don’t need the Dependabot PRs for each checkout bump.</p>
<p>Note that GitHub changed the behavior for actions a while ago to not allow short shas, but what you haven’t isn’t a short sha. I suppose it could have been a custom commit in a fork of actions/checkout that was deleted, but in general such things don’t get GCd…</p>

**Comment by GavinCameron** [comment, marked as answer]
<p>Hi jsoref. I think I’ve misunderstood something with actions. I’m having a d’oh Homer Simpson moment!</p>
<p>I thought that I checkout from my own repo where I created the action. It looks like I need to checkout actions from a universal, widely used by all repo <a href="https://github.com/actions/checkout/" class="inline-onebox" rel="noopener nofollow ugc">GitHub - actions/checkout: Action for checking out a repo</a>.</p>
<blockquote>
<p>I don’t know where this magic sha came from:</p>
</blockquote>
<p>Right! That’s the commit from my own repo at <code>&lt;ourorg&gt;/my-ssql-container-action</code> whereas thanks to your comment I now realize I need to use <a href="https://github.com/actions/checkout/commits/v2" rel="noopener nofollow ugc">this universally used actions repo</a> that you linked to.</p>
<p>I’m going to try that out today, thanks for the indirect tip</p>

**Comment by sikz1127** [comment]
<p>But if I use GithubActions and the secrets doesn’t work for me in order to connect with private key.  It is necessary to set a path but the secret has the key-value. How could this work?</p>
<p>env:<br>
SNOWSQL_RSA_PRIVATE_KEY: ${{ ecrets.SNOWSQL_RSA_PRIVATE_KEY }}</p>
<p>But when calling</p>
<p>~/bin/snowsql -q “use role $SNOWSQL_ROLE” -o friendly=false --private-key-path $SNOWSQL_RSA_PRIVATE_KEY</p>
<p>I got the error Path ‘’ does not exist "</p>
<p>Thanks in advance</p>

**Comment by GavinCameron** [comment]
<p><a class="mention" href="/sikz1127">@sikz1127</a> I don’t have an answer but can tell you what I did. Rather than refer to secrets in the Dockerfile, I instead used the entrypoint.<br>
e.g.</p>
<p>entrypoint.sh:</p>
<pre><code class="lang-auto">export SNOWSQL_RSA_PRIVATE_KEY=${{ Secrets.SNOWSQL_RSA_PRIVATE_KEY }}
</code></pre>
<p>Also note you have a typo:<br>
<code>${{ ecrets...</code> presumably should be <code>${{ Secrets...</code></p>

**Comment by sikz1127** [comment]
<p><a class="mention" href="/gavincameron">@GavinCameron</a>  Thank you for your reply, I will try it<br>
yes I have a copy/paste typo but the test were with the correct text :slight_smile:</p>
